rir/src/ir/Compiler.cpp

```diff
@@ -1072,6 +1072,150 @@ bool compileSpecialCall(CompilerContext& ctx, SEXP ast, SEXP fun, SEXP args_,
+         *   dup; asInteger(arg[0])
+         *   if (value == 1) br label[0]
+         *   ...
+         *   if (value == n) br label[n-1]
+         *   br nil
+         *
+         * str:
+         *   check if it fits in group[i], if yes jump to label[i]
+         *   if (value == group[0]) br labels[0]
+         *   ...
+         *   if (value == group[n-1]) br labels[n-1]
+         *   compile default arg (or br to nil if none)
+         *   br cont
+         *
+         * label[i]:
+         *   # if arg[i+1] is missing, we must came from integer case
+         *   # error("empty alternative in numeric switch")
+         *   compile expression[i]
+         *   br cont
+         * ...
+         *
+         * nil:
+         *   push R_NilValue
+         *
+         * cont:
+         *   # (stack is [1stArg, retval])
+         *   swap
+         *   pop # 1stArg
+         */
+        int argLen = args.length();
+        if (argLen <= 1) {
+            // this call will always give error/warning
+            // TODO add error instruction once available
+            return false;
+        }
+        // when 1st arg is string, switch behaves like C/C++ switch/case.
+        // Cases like `x=, y=, z=20` are grouped together.
+        // groups[-1] is not used, for impl convenience
+        std::vector<std::vector<SEXP>> groups = {{}};
+        std::vector<SEXP> expressions; // the ret-val ast for each group
+        std::vector<BC::Label> labels; // evaluate/return for each group
+        BC::Label vecEContBr = cs.mkLabel();
+        BC::Label facWContBr = cs.mkLabel();
+        BC::Label strBr  = cs.mkLabel();
+        BC::Label nilBr  = cs.mkLabel();
+        BC::Label contBr = cs.mkLabel();
+        SEXP dflt = nullptr; // ast for default ret-val
+        std::vector<bool> argMissing(argLen, true);
+
+        // find default and group args
+        int argIdx = 1;
+        for (RListIter arg = args.begin() + 1; arg != args.end(); ++arg, ++argIdx) {
+            if (!arg.hasTag()) {
+                if (dflt)
+                    // TODO multiple default with 1st arg type == string
+                    // i.e. compile `str` section with
+                    // error("duplicate 'switch' defaults")
+                    return false;
+                dflt = *arg;
+                continue;
+            }
+            groups.back().push_back(arg.tag());
+            if (*arg != R_MissingArg) {
+                expressions.push_back(*arg);
+                groups.push_back({});
+                labels.push_back(cs.mkLabel());
+                argMissing[argIdx] = false;
+            }
+        }
+        printf("compiling switch\n");
+        Rf_PrintValue(ast);
+
+        /******************* INSTRUCTIONS START HERE *********************/
+        compileExpr(ctx, args[0]);
+        // the inlined list of types is based on definition of `isVector` in `Rinlinedfuns.h`
+        for (auto type : {LGLSXP, INTSXP, REALSXP, CPLXSXP, STRSXP,
```

Comment:
I tried adding support for `BC::isType(TypeChecks::Vector)` (also for `String` and `Factor`), and handle it accordingly in `interp.cpp`. However I found that the `isType` instruction currently only comes from pir, so I decided to refrain from that and use `BC::is` instead.